doc/api/assert.md

```diff
@@ -2539,6 +2539,130 @@ assert.throws(throwingFirst, /Second$/);
 // AssertionError [ERR_ASSERTION]
 ```
 
+## `assert.matchObject(actual, expected[, message])`
+
+<!-- YAML
+added: REPLACEME
+-->
+
+* `actual` {any}
+* `expected` {any}
+* `message` {string|Error}
+
+Evaluates the equivalence between the `actual` and `expected` parameters by
+performing a deep comparison. This function ensures that all properties defined
+in the `expected` parameter exactly match those in the `actual` parameter in
+both value and type, without allowing type coercion.
+
+```mjs
+import assert from 'node:assert';
+
+assert.matchObject({ a: 1, b: '2' }, { a: 1, b: 2 });
+// OK
+
+assert.matchObject({ a: 1, b: '2', c: 3 }, { a: 1, b: 2 });
+// OK
+
+assert.matchObject({ a: { b: { c: '1' } } }, { a: { b: { c: 1 } } });
+// OK
+
+assert.matchObject({ a: 1 }, { a: 1, b: 2 });
+// AssertionError
+
+assert.matchObject({ a: 1, b: true }, { a: 1, b: 'true' });
+// AssertionError
+
+assert.matchObject({ a: { b: 2 } }, { a: { b: 2, c: 3 } });
+// AssertionError
+```
+
+```cjs
+const assert = require('node:assert');
+
+assert.matchObject({ a: 1, b: '2' }, { a: 1, b: 2 });
+// OK
+
+assert.matchObject({ a: 1, b: '2', c: 3 }, { a: 1, b: 2 });
+// OK
+
+assert.matchObject({ a: { b: { c: '1' } } }, { a: { b: { c: 1 } } });
+// OK
+
+assert.matchObject({ a: 1 }, { a: 1, b: 2 });
+// AssertionError: Expected key b
+
+assert.matchObject({ a: 1, b: true }, { a: 1, b: 'true' });
+// AssertionError
+
+assert.matchObject({ a: { b: 2, d: 4 } }, { a: { b: 2, c: 3 } });
+// AssertionError: Expected key c
+```
+
+If the values or keys are not equal in the `expected` parameter, an [`AssertionError`][] is thrown with a `message`
+property set equal to the value of the `message` parameter. If the `message`
+parameter is undefined, a default error message is assigned. If the `message`
+parameter is an instance of an [`Error`][] then it will be thrown instead of the
+`AssertionError`.
+
+## `assert.matchObjectStrict(actual, expected[, message])`
+
+<!-- YAML
+added: REPLACEME
+-->
+
+* `actual` {any}
+* `expected` {any}
+* `message` {string|Error}
+
+Assesses the equivalence between the `actual` and `expected` parameters through a
+deep comparison, ensuring that all properties in the `expected` parameter are
+present in the `actual` parameter with equivalent values, permitting type coercion
+where necessary.
+
+```mjs
+import assert from 'node:assert';
+
+assert.matchObject({ a: 1, b: 2 }, { a: 1, b: 2 });
+// OK
+
+assert.matchObject({ a: { b: { c: 1 } } }, { a: { b: { c: 1 } } });
+// OK
+
+assert.matchObject({ a: 1, b: 2, c: 3 }, { a: 1, b: 2 });
+// OK
```

Comment:
I'm still not convinced that this is sufficiently different from `deepEqual` and `deepStrictEqual` to warrant a new API. At the absolute least the documentation does not provide enough information for someone to be able to decide which to use

Reply:
There was some conversation in the original issue about whether it should be:

1. a new method, e.g. matchObject
2. an option to the existing method, e.g.  deepStrictEqual(..., { partial: tru })
3. a utility function for the existing method, e.g. deepStrictEqual(..., objectContaining(...))

here @synapse decided to go with option 1, lacking a clear consensus on which option made the most sense

Reply:
> I'm still not convinced that this is sufficiently different from `deepEqual` and `deepStrictEqual` to warrant a new API. At the absolute least the documentation does not provide enough information for someone to be able to decide which to use

I agree. I would love to see `node:assert` have a `match` function equivalent to https://github.com/tapjs/tapjs/blob/8f2baa7eb5e1f4cd0fcf8b60da4fb4e41ea680f1/docs/src/content/docs/api/asserts/index.md?plain=1#L318-L338

But the proposal here seems like a less consistent `deepStrictEqual`.